docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1058 @@
+* `DOTNET_EnableHWIntrinsic=0`
+
+### Benchmarking
+
+All that complexity needs to pay off. We need to **benchmark the code to verify that the investment is beneficial**. We can do that with [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet).
+
+#### Custom config
+
+It's possible to define a config that instructs the harness to run the benchmarks for all four scenarios:
+
+```cs
+static void Main(string[] args)
+{
+    Job enough = Job.Default
+        .WithWarmupCount(1)
+        .WithIterationTime(TimeInterval.FromSeconds(0.25))
+        .WithMaxIterationCount(20);
+
+    IConfig config = DefaultConfig.Instance
+        .HideColumns(Column.EnvironmentVariables, Column.RatioSD, Column.Error)
+        .AddDiagnoser(new DisassemblyDiagnoser(new DisassemblyDiagnoserConfig
+            (exportGithubMarkdown: true, printInstructionAddresses: false)))
+        .AddJob(enough.WithEnvironmentVariable("DOTNET_EnableHWIntrinsic", "0").WithId("Scalar").AsBaseline());
+
+    if (Vector256.IsHardwareAccelerated)
+    {
+        config = config
+            .AddJob(enough.WithId("Vector256"))
+            .AddJob(enough.WithEnvironmentVariable("DOTNET_EnableAVX2", "0").WithId("Vector128"));
+
+    }
+    else if (Vector128.IsHardwareAccelerated)
+    {
+        config = config.AddJob(enough.WithId("Vector128"));
+    }
+
+    BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly)
+        .Run(args, config);
+}
+```
+
+**Note:** the config defines a [disassembler](https://adamsitnik.com/Disassembly-Diagnoser/), which exports a disassembly in GitHub markdown format (supported on both x64 and arm64, Windows and Linux). It is very often an invaluable tool when working with high-performance code where inspecting generated assembly code is required.
+
+#### Memory alignment
+
+BenchmarkDotNet does a lot of heavy lifting for the end users, but it can not protect us from the random memory alignment which can be different per each benchmark run and affect the stability of the benchmarks.
+
+We have three possibilities:
+
+* We can enforce the alignment ourselves and have very stable results.
+* We can ask the harness to try to randomize the memory and observe entire possible distribution with each run.
+* We can do nothing and wonder why the results vary from time to time.
+
+##### Enforcing memory alignment
+
+We can allocate aligned unmanaged memory by using the [NativeMemory.AlignedAlloc](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativememory.alignedalloc).
+
+```cs
+public unsafe class Benchmarks
+{
+    private void* _pointer;
+
+    [Params(6, 32, 1024)] // test various sizes
+    public uint Size;
+
+    [GlobalSetup]
+    public void Setup()
+    {
+        _pointer = NativeMemory.AlignedAlloc(byteCount: Size * sizeof(int), alignment: 32);
+        new Span<int>(_pointer, (int)Size).Fill(0); // ensure it's all zeros, so 1 is never found
```

Comment:
Perhaps `NativeMemory.Clear` would be the best choice?

```suggestion
        NativeMemory.Clear(_pointer, byteCount: Size * sizeof(int)); // ensure it's all zeros, so 1 is never found
```

Reply:
Yes it is.
(I wasn't aware of that method, so thanks!)

Reply:
Filed https://github.com/dotnet/runtime/issues/84126, I hoped there would not be any difference 🙂